### ISSUE TITLE
java.lang.OutOfMemoryError:com.intercom.input.gallery.AnnotatedImageView.createCanvasWithSameBoundsAsRenderedDrawable

### ISSUE THREAD

**Issue by pengyuan1990** [intercom/intercom-android#371]
## Version info
  - Intercom for Android version:4.3
  - Android OS version:18
  - Android Device:samsung:GT-I9308:armeabi-v7a

## Expected behavior
no crash

## Actual behavior
crash

## Steps to reproduce
 1. open intercom
 2. send image 

## Logs

java.lang.OutOfMemoryError
at android.graphics.Bitmap.nativeCreate(Bitmap.java)
at android.graphics.Bitmap.createBitmap(Bitmap.java:726)
at android.graphics.Bitmap.createBitmap(Bitmap.java:703)
at android.graphics.Bitmap.createBitmap(Bitmap.java:670)
atcom.intercom.input.gallery.AnnotatedImageView.createCanvasWithSameBoundsAsRenderedDrawable(AnnotatedImageView.java:128)
at com.intercom.input.gallery.AnnotatedImageView.access$000(AnnotatedImageView.java:20)
at com.intercom.input.gallery.AnnotatedImageView$1.run(AnnotatedImageView.java:86)
at android.os.Handler.handleCallback(Handler.java:730)
at android.os.Handler.dispatchMessage(Handler.java:92)
at android.os.Looper.loop(Looper.java:176)
at android.app.ActivityThread.main(ActivityThread.java:5454)
at java.lang.reflect.Method.invokeNative(Method.java)
at java.lang.reflect.Method.invoke(Method.java:525)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1209)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1025)
at dalvik.system.NativeStart.main(NativeStart.java)

**Comment by ronocod** [comment]
Hi, we're investigating this now.

**Comment by apperside** [comment]
Hi, same issue here, here the stacktrace

```
Fatal Exception: java.lang.OutOfMemoryError: Failed to allocate a 10227216 byte allocation with 5293641 free bytes and 5MB until OOM
       at dalvik.system.VMRuntime.newNonMovableArray(VMRuntime.java)
       at android.graphics.Bitmap.nativeCreate(Bitmap.java)
       at android.graphics.Bitmap.createBitmap(Bitmap.java:939)
       at android.graphics.Bitmap.createBitmap(Bitmap.java:912)
       at android.graphics.Bitmap.createBitmap(Bitmap.java:879)
       at com.intercom.input.gallery.AnnotatedImageView.createCanvasWithSameBoundsAsRenderedDrawable(AnnotatedImageView.java:128)
       at com.intercom.input.gallery.AnnotatedImageView.access$000(AnnotatedImageView.java:20)
       at com.intercom.input.gallery.AnnotatedImageView$1.run(AnnotatedImageView.java:86)
       at android.os.Handler.handleCallback(Handler.java:739)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:145)
       at android.app.ActivityThread.main(ActivityThread.java:5938)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1400)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1195)
```

**Comment by noinlijin** [comment]
```
java.lang.OutOfMemoryError
at android.graphics.Bitmap.nativeCreate(Native Method)
at android.graphics.Bitmap.createBitmap(Bitmap.java:811)
at android.graphics.Bitmap.createBitmap(Bitmap.java:788)
at android.graphics.Bitmap.createBitmap(Bitmap.java:755)
at com.intercom.input.gallery.AnnotatedImageView.d(Unknown Source)
at com.intercom.input.gallery.AnnotatedImageView.a(Unknown Source)
at com.intercom.input.gallery.AnnotatedImageView$1.run(Unknown Source)
at android.os.Handler.handleCallback(Handler.java:733)
at android.os.Handler.dispatchMessage(Handler.java:95)
at android.os.Looper.loop(Looper.java:136)
at android.app.ActivityThread.main(ActivityThread.java:5377)
at java.lang.reflect.Method.invokeNative(Native Method)
at java.lang.reflect.Method.invoke(Method.java:515)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:976)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:792)
at dalvik.system.NativeStart.main(Native Method)
```

this also very easy to happen in 30 device of I tested every time

**Comment by noinlijin** [comment]
please fix this bug quickly  
![image](https://user-images.githubusercontent.com/3073370/28996156-ec282de2-7a2c-11e7-93a9-7543fc90f08c.png)

**Comment by noinlijin** [comment]
there were some reference to solve this bug
详细说明
问题分析：

OutOfMemoryError是在当程序需要请求一块内存，而java虚拟机进行垃圾回收机制GC后无法再提供要求的内存时抛出此异常。

解决方案：

OutOfMemoryError是在当程序需要请求一块内存，而java虚拟机进行垃圾回收机制GC后无法再提供要求的内存时抛出此类异常，程序已经占用的内存到达系统限制的应用内存上线无法再申请到所需的内存空间。需要重点检查代码中是否有死循环或递归调用。检查是否在循环重复产生占用大内存的新对象。是否有一次获得大量数据的查询存到内存中，应当尽量用分页的方式查询。是否List、Map等集合对象使用后未清除，而集合对象有对对象的引用导致对象不能被GC回收，是否有内存泄露等。建议使用深度性能测试检测内存泄漏等问题。

代码示例：

    // 在循环外初始化大对象list并在循环内重复使用，并清空集合类、置null
    List list=new ArrayList<Object>();
    for(Object tmp:otherList){
      // to do something
    } 
    list.clear();
    list=null;


参考文献：

1.https://developer.android.com/reference/java/lang/OutOfMemoryError.html

2.https://source.android.com/compatibility/android-cdd.html#7_6_memory_and_storage （安卓系统对程序所使用内存的最大值限制，参见3.7节的表格，或直接参考文献3)

3.https://drive.google.com/file/d/0B7Vx1OvzrLa3Y0R0X1BZbUpicGc/view?usp=sharing（安卓系统对内存最大值限制）

**Comment by noinlijin** [comment]
another crash occur like @apperside , I have more info 

device info: 三星 SM-A5000 5.0.2 720x1280 高通 骁龙410



```   
java.lang.OutOfMemoryError: Failed to allocate a 4545436 byte allocation with 1625615 free bytes and 1587KB until OOM
at dalvik.system.VMRuntime.newNonMovableArray(Native Method)
at android.graphics.Bitmap.nativeCreate(Native Method)
at android.graphics.Bitmap.createBitmap(Bitmap.java:947)
at android.graphics.Bitmap.createBitmap(Bitmap.java:918)
at android.graphics.Bitmap.createBitmap(Bitmap.java:885)
at com.intercom.input.gallery.AnnotatedImageView.d(Unknown Source)
at com.intercom.input.gallery.AnnotatedImageView.a(Unknown Source)
at com.intercom.input.gallery.AnnotatedImageView$1.run(Unknown Source)
at android.os.Handler.handleCallback(Handler.java:739)
at android.os.Handler.dispatchMessage(Handler.java:95)
at android.os.Looper.loop(Looper.java:145)
at android.app.ActivityThread.main(ActivityThread.java:6145)
at java.lang.reflect.Method.invoke(Native Method)
at java.lang.reflect.Method.invoke(Method.java:372)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1399)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1194)
```

![image](https://user-images.githubusercontent.com/3073370/28996197-c2810490-7a2d-11e7-97a2-86fdeeb3974a.png)

and device: 三星 SM-G6100 6.0.1 1080x1920 高通 骁龙625

```
java.lang.OutOfMemoryError: Failed to allocate a 10784668 byte allocation with 10188000 free bytes and 9MB until OOM
at dalvik.system.VMRuntime.newNonMovableArray(Native Method)
at android.graphics.Bitmap.nativeCreate(Native Method)
at android.graphics.Bitmap.createBitmap(Bitmap.java:975)
at android.graphics.Bitmap.createBitmap(Bitmap.java:946)
at android.graphics.Bitmap.createBitmap(Bitmap.java:913)
at com.intercom.input.gallery.AnnotatedImageView.d(Unknown Source)
at com.intercom.input.gallery.AnnotatedImageView.a(Unknown Source)
at com.intercom.input.gallery.AnnotatedImageView$1.run(Unknown Source)
at android.os.Handler.handleCallback(Handler.java:739)
at android.os.Handler.dispatchMessage(Handler.java:95)
at android.os.Looper.loop(Looper.java:148)
at android.app.ActivityThread.main(ActivityThread.java:7325)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1230)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1120)
```

**Comment by noinlijin** [comment]
the intercom code version is : 3.2.8

**Comment by niteshverma09** [comment]
I am facing the same issue at `com.intercom.input.gallery.AnnotatedImageView.createCanvasWithSameBoundsAsRenderedDrawable (SourceFile:129)`. when can we expect a fix ??

`Fatal Exception: java.lang.OutOfMemoryError
       at android.graphics.Bitmap.nativeCreate(Bitmap.java)
       at android.graphics.Bitmap.createBitmap(Bitmap.java:669)
       at android.graphics.Bitmap.createBitmap(Bitmap.java:649)
       at com.intercom.input.gallery.AnnotatedImageView.createCanvasWithSameBoundsAsRenderedDrawable(SourceFile:129)
       at com.intercom.input.gallery.AnnotatedImageView.access$000(SourceFile:21)
       at com.intercom.input.gallery.AnnotatedImageView$1.run(SourceFile:87)
       at android.os.Handler.handleCallback(Handler.java:615)
       at android.os.Handler.dispatchMessage(Handler.java:92)
       at android.os.Looper.loop(Looper.java:137)
       at android.app.ActivityThread.main(ActivityThread.java:4960)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:511)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1038)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:805)
       at dalvik.system.NativeStart.main(NativeStart.java)`

**Comment by ronocod** [comment]
We have a fix merged, we'll release it later this morning.

**Comment by DaleCantwell** [comment]
The fix is available now in [4.0.0](https://github.com/intercom/intercom-android/releases/tag/4.0.0)
